cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -229,16 +245,29 @@ private ProgramOptions createPluginSnapshot(ProgramOptions options, ProgramId pr
     builder.putAll(options.getArguments().asMap());
     for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
       Plugin plugin = pluginEntry.getValue();
-      File destFile = new File(tempDir, Artifacts.getFileName(plugin.getArtifactId()));
+      final File destFile = new File(tempDir, Artifacts.getFileName(plugin.getArtifactId()));
       // Skip if the file has already been copied.
       if (!files.add(destFile.getName())) {
         continue;
       }
 
       try {
         ArtifactId artifactId = Artifacts.toArtifactId(programId.getNamespaceId(), plugin.getArtifactId());
-        ArtifactDetail detail = artifactRepository.getArtifact(artifactId.toId());
-        Locations.linkOrCopy(detail.getDescriptor().getLocation(), destFile);
+        final ArtifactDetail detail = artifactRepository.getArtifact(artifactId.toId());
+        try {
+          new NamespacedImpersonator(artifactId.getParent(), impersonator).impersonate(
```

Comment:
can't we use regular impersonator here?
And several other places?
It only makes sense to use NamespacedImpersonator if you're going to reuse the NamespacedImpersonator more than once per creation, right?